shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,313 @@
+     */
+    template <typename T>
+    struct Entering : public ProviderBase<T, commands::ProviderClassification::ENTERING> {};
+
+    /**
+     * This DSL word is used to define a leaving transition from a Provider.
+     * It should normally be combined with a Causing DSL word.
+     * In this case when another Provider needs control and uses a When condition, this will run until that
+     * condition is fulfilled. A Leaving condition will be preferred over an Entering condition that provides the
+     * same Causing statement (the system will rather a module cleans up after itself than ask another module to do
+     * so).
+     *
+     * @tparam T the Provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Leaving : public ProviderBase<T, commands::ProviderClassification::LEAVING> {};
+
+    /**
+     * This DSL word is used to define a Provider for a type.
+     * It will execute in the same way as an on<Trigger<T>> statement except that when it executes will be
+     * determined by the Director. This ensures that it will only run when it has permission to run and will
+     * transition between Providers in a sensible way.
+     *
+     * @tparam T the Provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Provides : public ProviderBase<T, commands::ProviderClassification::NORMAL> {};
+
+    /**
+     * The When DSL word is used to limit access to a Provider unless a condition is true.
+     * This will prevent a Provider from running if the state is false.
+     * However if there is a Entering<T> or Leaving reaction that has a Causing relationship for this When, it will
+     * be executed to ensure that the Provider can run.
+     *
+     * @tparam T the condition expression that must be true in order to execute this task.
+     */
+    template <typename State, template <typename> class expr, enum State::Value value>
+    struct When {
+
+        /**
+         * Bind a when expression in the Director.
+         *
+         * @tparam DSL the DSL from NUClear
+         *
+         * @param reaction the reaction that is having this when condition bound to it
+         */
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the director about this when condition
+            reaction->reactor.powerplant.emit(std::make_unique<commands::WhenExpression>(
+                reaction,
+                // typeindex of the enum value
+                typeid(State),
+                // Function that uses expr to determine if the passed value v is valid
+                [](const int& v) { return expr<int>()(v, value); },
+                // Function that uses get to get the current state of the reaction
+                []() { return NUClear::dsl::operation::CacheGet<State>::get(); },
```

Comment:
```suggestion
                [](const int& v) -> bool { return expr<int>()(v, value); },
                // Function that uses get to get the current state of the reaction
                []() -> int { return NUClear::dsl::operation::CacheGet<State>::get(); },
```

Reply:
Why put the return types in?

Reply:
I mean I guess sure I can put them in anyway...
Also you have drawn my attention to the fact that that get is wrong as it returns a shared_ptr

Reply:
consistency (with the lambda two lines down), and the raw function pointers you are using (in `WhenExpression`) explicitly specify what the return type should be, so why not?